homeassistant/components/sonos/speaker.py

```diff
@@ -217,3 +244,57 @@ async def async_unseen(self, now: datetime.datetime | None = None) -> None:
             await subscription.unsubscribe()
 
         self._subscriptions = []
+
+    async def async_update_device_properties(self, event: SonosEvent = None) -> None:
+        """Update device properties using the provided SonosEvent."""
+        if event is None:
+            return
+
+        if (more_info := event.variables.get("more_info")) is not None:
+            battery_dict = dict(x.split(":") for x in more_info.split(","))
+            await self.async_update_battery_info(battery_dict)
+
+        self.async_write_entity_states()
+
+    async def async_update_battery_info(self, battery_dict: dict[str, Any]) -> None:
+        """Update battery info using the decoded SonosEvent."""
+        self._last_battery_event = dt_util.utcnow()
+
+        is_charging = EVENT_CHARGING[battery_dict["BattChg"]]
+        if is_charging == self.charging:
+            self.battery_info.update({"Level": int(battery_dict["BattPct"])})
+        else:
+            if battery_info := await self.hass.async_add_executor_job(
+                fetch_battery_info_or_none, self.soco
+            ):
+                self.battery_info = battery_info
+
+    @property
+    def power_source(self) -> str:
+        """Return the name of the current power source.
+
+        Observed to be either BATTERY or SONOS_CHARGING_RING or USB_POWER.
+        """
+        return self.battery_info.get("PowerSource", STATE_UNKNOWN)
```

Comment:
If we don't know the value we don't return the item at all in the state attributes.

Reply:
This can only protect from a missing `PowerSource` key on a device which previously provided one. This doesn't seem like it should ever occur from my testing, so probably unnecessary to have a fallback.